File: services/activities-core/src/ws/storyBuilder.ts
import { SocketStream } from '@fastify/websocket';
import { FastifyRequest } from 'fastify';
import { recordGameResult } from '../services/stats';

export interface StoryParagraph {
    userId: string;
    text: string;
    votes: Record<string, number>; // userId -> score (0-10)
}

type Gender = 'boy' | 'girl';

// Story prompts to give players inspiration
const GENERIC_PROMPTS = [
    { title: "The Mystery Box", opening: "The package arrived without a return address. Inside was a single, glowing object..." },
    { title: "The Time Machine", opening: "It looked like a normal watch, but when they turned the dial, the world around them shifted..." },
    { title: "The Haunted House", opening: "Everyone said the old mansion was abandoned. But the lights in the attic window told a different story..." },
];

const ROMANCE_PROMPTS = [
    { title: "Coffee Shop Meet-Cute", opening: "The morning rush at the campus coffee shop was chaos as usual. That's when their eyes met across the crowded counter..." },
    { title: "Rainy Day Encounter", opening: "The sudden downpour caught everyone off guard. Two strangers found themselves sharing the same tiny awning..." },
    { title: "The Wrong Order", opening: "'I think you grabbed my food by mistake,' they said, holding up an identical takeout bag. Their smile was unexpectedly warm..." },
    { title: "Study Group Sparks", opening: "The study group was supposed to be about calculus. But every time their hands brushed reaching for the same textbook..." },
    { title: "Elevator Stuck", opening: "The elevator lurched to a stop between floors. 'Well,' they said to the only other person inside, 'I guess we have time to talk...'" },
    { title: "The Time Capsule", opening: "They dug up a time capsule from 50 years ago. Inside was a photo of two people who looked exactly like them..." },
    { title: "The Last Train", opening: "Missing the last train home seemed like a disaster, until they realized who else was stranded on the platform..." },
];

const GAY_ROMANCE_PROMPTS = [
    { title: "Locker Room Confession", opening: "After practice, the locker room was empty. He thought he was alone until he saw him..." },
    { title: "The Late Night Gamer", opening: "They had been gaming together online for months. When they finally met at the convention, the chemistry was instant..." },
    { title: "Roommates", opening: "They had been roommates for a year, strictly platonic. But one movie night changed everything..." },
    { title: "The Rivalry", opening: "They were academic rivals, always competing for the top spot. But beneath the competition was something else..." },
    { title: "The Fake Date", opening: "He needed a date for his sister's wedding to get his family off his back. His best friend volunteered, but the lines started to blur..." },
    { title: "The Coffee Shop Rivalry", opening: "He owned the artisanal tea shop. The guy across the street owned the espresso bar. It was war, until the power went out on the whole block..." },
];

const LESBIAN_ROMANCE_PROMPTS = [
    { title: "The Bookstore Corner", opening: "She reached for the same book on the top shelf. Their fingers touched, and time seemed to stop..." },
    { title: "Art Class Model", opening: "She was trying to focus on her sketch, but the model's gaze was distracting her in the best way possible..." },
    { title: "The Coffee Date", opening: "It was supposed to be just a friendly coffee. Three hours later, neither of them wanted to leave..." },
    { title: "Softball Season", opening: "The game was intense, but she couldn't take her eyes off the pitcher..." },
    { title: "The Road Trip", opening: "Her car broke down in the middle of nowhere. The mechanic who pulled up in the tow truck was the most beautiful woman she'd ever seen..." },
    { title: "The Masquerade", opening: "She didn't know who she was dancing with behind the mask, but she knew she never wanted the music to stop..." },
];

function getPromptForGenders(genders: Gender[]) {
    if (genders.length !== 2) return GENERIC_PROMPTS[Math.floor(Math.random() * GENERIC_PROMPTS.length)];

    const [g1, g2] = genders;

    if (g1 === 'boy' && g2 === 'boy') {
        return GAY_ROMANCE_PROMPTS[Math.floor(Math.random() * GAY_ROMANCE_PROMPTS.length)];
    } else if (g1 === 'girl' && g2 === 'girl') {
        return LESBIAN_ROMANCE_PROMPTS[Math.floor(Math.random() * LESBIAN_ROMANCE_PROMPTS.length)];
    } else {
        // Boy + Girl or Girl + Boy
        return ROMANCE_PROMPTS[Math.floor(Math.random() * ROMANCE_PROMPTS.length)];
    }
}

export interface StoryBuilderSession {
    id: string;
    activityKey: 'story_builder';
    status: 'pending' | 'countdown' | 'writing' | 'voting' | 'ended';
    phase: 'lobby' | 'countdown' | 'writing' | 'voting' | 'ended';
    lobbyReady: boolean;
    creatorUserId: string;
    participants: Array<{ userId: string; joined: boolean; ready: boolean; score: number; gender?: Gender }>;
    createdAt: number;
    roundStartedAt?: number;
    paragraphs: StoryParagraph[];
    maxParagraphsPerUser: number;
    currentTurnUserId?: string;
    turnOrder: string[];
    turnIndex: number;
    winnerUserId?: string | null;
    storyPrompt?: { title: string; opening: string };
    leaveReason?: 'opponent_left' | 'forfeit' | null;
    statsRecorded?: boolean;  // Guard against duplicate stat recording
}

const sessions: Record<string, StoryBuilderSession> = {};
const connections: Record<string, Set<any>> = {};
const userSockets: Record<string, Map<string, any>> = {}; // sessionId -> userId -> socket

// Session cleanup configuration (prevents memory leaks)
const SESSION_CLEANUP_INTERVAL_MS = 5 * 60 * 1000; // 5 minutes
const SESSION_ENDED_TTL_MS = 60 * 60 * 1000; // 1 hour after ending
const SESSION_PENDING_TTL_MS = 24 * 60 * 60 * 1000; // 24 hours if never started

/**
 * Cleanup stale sessions to prevent memory leaks.
 */
function cleanupStaleSessions(): void {
    const now = Date.now();
    let cleanedCount = 0;

    for (const sessionId of Object.keys(sessions)) {
        const session = sessions[sessionId];
        if (!session) continue;

        const age = now - session.createdAt;
        const shouldClean =
            (session.status === 'ended' && age > SESSION_ENDED_TTL_MS) ||
            (session.status === 'pending' && age > SESSION_PENDING_TTL_MS);

        if (shouldClean) {
            delete sessions[sessionId];
            delete connections[sessionId];
            delete userSockets[sessionId];
            cleanedCount++;
        }
    }

    if (cleanedCount > 0) {
        console.log(`[StoryBuilder] Cleaned up ${cleanedCount} stale sessions. Active: ${Object.keys(sessions).length}`);
    }
}

// Start cleanup interval
setInterval(cleanupStaleSessions, SESSION_CLEANUP_INTERVAL_MS);

const MAX_PARAGRAPHS_PER_USER = 3;

export function createStoryBuilderSession(creatorUserId: string, participants: string[] = [], id?: string): string {
    const sessionId = id || `sb-${Math.random().toString(36).substring(2, 10)}`;
    const uniqueParticipants = Array.from(new Set([creatorUserId, ...participants]));

    sessions[sessionId] = {
        id: sessionId,
        activityKey: 'story_builder',
        status: 'pending',
        phase: 'lobby',
        lobbyReady: false,
        creatorUserId,
        participants: uniqueParticipants.map(uid => ({ userId: uid, joined: uid === creatorUserId, ready: false, score: 0 })),
        createdAt: Date.now(),
        paragraphs: [],
        maxParagraphsPerUser: MAX_PARAGRAPHS_PER_USER,
        turnOrder: [],
        turnIndex: 0,
        winnerUserId: null,
        storyPrompt: undefined // Will be set when game starts based on genders
    };
    connections[sessionId] = new Set();
    return sessionId;
}

export function getStoryBuilderSession(sessionId: string): StoryBuilderSession | undefined {
    return sessions[sessionId];
}

export function listStoryBuilderSessions() {
    return Object.values(sessions);
}

export function joinStoryBuilder(sessionId: string, userId: string, socket?: any) {
    const session = sessions[sessionId];
    if (!session) throw new Error('session_not_found');

    const participant = session.participants.find(p => p.userId === userId);
    if (participant) {
        participant.joined = true;
    } else {
        session.participants.push({ userId, joined: true, ready: false, score: 0 });
    }

    // Track user's socket for disconnect detection
    if (socket) {
        if (!userSockets[sessionId]) userSockets[sessionId] = new Map();
        userSockets[sessionId].set(userId, socket);
    }

    session.lobbyReady = session.participants.every(p => p.ready) && session.participants.length >= 2;
    broadcastState(sessionId);
    return session;
}

export function setStoryBuilderReady(sessionId: string, userId: string, ready: boolean, gender?: Gender) {
    const session = sessions[sessionId];
    if (!session) throw new Error('session_not_found');

    const participant = session.participants.find(p => p.userId === userId);
    if (participant) {
        participant.ready = ready;
        if (gender) {
            participant.gender = gender;
        }
    }

    session.lobbyReady = session.participants.every(p => p.ready) && session.participants.length >= 2;

    if (session.lobbyReady && session.status === 'pending') {
        // Generate prompt based on genders
        const genders = session.participants.map(p => p.gender).filter((g): g is Gender => !!g);
        session.storyPrompt = getPromptForGenders(genders);

        startCountdown(sessionId);
    } else {
        broadcastState(sessionId);
    }
    return session;
}

function startCountdown(sessionId: string) {
    const session = sessions[sessionId];
    if (!session) return;

    session.status = 'countdown';
    session.phase = 'countdown';
    broadcastState(sessionId);

    setTimeout(() => {
        startStory(sessionId);
    }, 5000);
}

export function startStory(sessionId: string) {
    const session = sessions[sessionId];
    if (!session) return;

    session.status = 'writing';
    session.phase = 'writing';
    session.roundStartedAt = Date.now();
    // Randomize turn order
    session.turnOrder = session.participants.map(p => p.userId).sort(() => Math.random() - 0.5);
    session.turnIndex = 0;
    session.currentTurnUserId = session.turnOrder[0];

    broadcastState(sessionId);
}

// Leave/forfeit handler
export function leaveStoryBuilder(sessionId: string, userId: string): { sessionEnded: boolean; winnerUserId?: string } {
    const session = sessions[sessionId];
    if (!session) throw new Error('session_not_found');

    // Remove user from user-socket tracking
    userSockets[sessionId]?.delete(userId);

    // Remove from participants
    const idx = session.participants.findIndex(p => p.userId === userId);
    if (idx !== -1) {
        session.participants.splice(idx, 1);
    }

    // If game was in progress (writing or voting), forfeit - remaining player wins
    if (session.status === 'writing' || session.status === 'voting' || session.status === 'countdown') {
        const remaining = session.participants.filter(p => p.joined);
        if (remaining.length === 1) {
            // Award win to remaining player
            const winner = remaining[0];
            winner.score = Math.max(winner.score, 100); // Forfeit win bonus
            session.winnerUserId = winner.userId;
            session.status = 'ended';
            session.phase = 'ended';
            session.leaveReason = 'opponent_left';

            // Record stats using fixed points (only if not already recorded)
            if (!session.statsRecorded) {
                session.statsRecorded = true;
                (async () => {
                    try {
                        await recordGameResult(winner.userId, 'story_builder', 'win', 200);  // Fixed: 50 + 150
                        await recordGameResult(userId, 'story_builder', 'loss', 50);  // Fixed: 50
                    } catch (err) {
                        console.error('[StoryBuilder] Failed to record game stats (forfeit):', err);
                    }
                })();
            }

            broadcastState(sessionId);
            return { sessionEnded: true, winnerUserId: winner.userId };
        } else if (remaining.length === 0) {
            // No one left, just end
            session.status = 'ended';
            session.phase = 'ended';
            session.leaveReason = 'opponent_left';
            broadcastState(sessionId);
            return { sessionEnded: true };
        }
    }

    // Lobby phase - just remove and update
    session.lobbyReady = session.participants.every(p => p.ready) && session.participants.length >= 2;

    // If we're in lobby phase and not enough players remain, end the session
    if (session.status === 'pending' && session.participants.length < 2) {
        session.status = 'ended';
        session.phase = 'ended';
        session.leaveReason = 'opponent_left';
        broadcastState(sessionId);
        return { sessionEnded: true };
    }

    broadcastState(sessionId);
    return { sessionEnded: false };
}

// Handle disconnect (called when socket closes)
function handleDisconnect(sessionId: string, socket: any) {
    // Find which user this socket belonged to
    const userMap = userSockets[sessionId];
    if (!userMap) return;

    let disconnectedUserId: string | null = null;
    for (const [userId, sock] of userMap.entries()) {
        if (sock === socket) {
            disconnectedUserId = userId;
            break;
        }
    }

    if (disconnectedUserId) {
        console.log(`[StoryBuilder] User ${disconnectedUserId} disconnected from session ${sessionId}`);
        leaveStoryBuilder(sessionId, disconnectedUserId);
    }
}

export function handleStoryBuilderConnection(connection: SocketStream, _req: FastifyRequest, sessionId: string) {
    const socket = connection.socket;
    if (!sessions[sessionId]) return;

    if (!connections[sessionId]) connections[sessionId] = new Set();
    connections[sessionId].add(socket);

    // Track this socket temporarily until we know the userId from join message
    let connectedUserId: string | null = null;

    socket.send(JSON.stringify({ type: 'state', payload: sessions[sessionId] }));

    socket.on('message', (message: Buffer) => {
        try {
            const data = JSON.parse(message.toString());
            // Capture userId from join message
            if (data.type === 'join' && data.payload?.userId) {
                connectedUserId = data.payload.userId;
                if (!userSockets[sessionId]) userSockets[sessionId] = new Map();
                userSockets[sessionId].set(connectedUserId!, socket);
            }
            handleMessage(sessionId, data, socket);
        } catch (e) {
            console.error('Failed to parse message', e);
        }
    });

    socket.on('close', () => {
        connections[sessionId]?.delete(socket);
        // Handle disconnect for forfeit logic
        if (connectedUserId) {
            handleDisconnect(sessionId, socket);
        }
    });
}

export function handleMessage(sessionId: string, data: any, socket?: any) {
    const session = sessions[sessionId];
    if (!session) return;

    if (data.type === 'join') {
        const { userId } = data.payload;
        if (userId) {
            joinStoryBuilder(sessionId, userId, socket);
        }
    } else if (data.type === 'leave') {
        const { userId } = data.payload;
        if (userId) {
            leaveStoryBuilder(sessionId, userId);
        }
    } else if (data.type === 'ready') {
        const { userId, gender } = data.payload;
        const participant = session.participants.find(p => p.userId === userId);
        if (participant) {
            setStoryBuilderReady(sessionId, userId, !participant.ready, gender);
        }
    } else if (data.type === 'submit_paragraph') {
        const { userId, text } = data.payload;

        // Validate turn
        if (session.status !== 'writing') return;
        if (session.currentTurnUserId !== userId) return;

        // Add paragraph
        session.paragraphs.push({
            userId,
            text,
            votes: {}
        });

        // Check if game should end or move to next turn
        const totalParagraphs = session.paragraphs.length;
        const maxTotal = session.participants.length * session.maxParagraphsPerUser;

        if (totalParagraphs >= maxTotal) {
            session.status = 'voting';
            session.phase = 'voting';
            session.currentTurnUserId = undefined;
        } else {
            // Next turn
            session.turnIndex = (session.turnIndex + 1) % session.participants.length;
            session.currentTurnUserId = session.turnOrder[session.turnIndex];
        }
        broadcastState(sessionId);

    } else if (data.type === 'vote_paragraph') {
        const { userId, paragraphIndex, score } = data.payload;

        if (session.status !== 'voting') return;
        if (score < 0 || score > 10) return;

        const paragraph = session.paragraphs[paragraphIndex];
        if (!paragraph) return;

        // Prevent voting on own paragraph? The prompt says "i canvote on my frien'd paragraph". 
        // Usually you can't vote on your own. Let's enforce that.
        if (paragraph.userId === userId) return;

        paragraph.votes[userId] = score;

        checkVotingComplete(sessionId);
        broadcastState(sessionId);
    }
}

function checkVotingComplete(sessionId: string) {
    const session = sessions[sessionId];
    if (!session) return;

    // Check if every paragraph has votes from all other participants
    const allVoted = session.paragraphs.every(p => {
        const otherParticipants = session.participants.filter(part => part.userId !== p.userId);
        return otherParticipants.every(part => p.votes[part.userId] !== undefined);
    });

    if (allVoted) {
        calculateScores(sessionId);
    }
}

function calculateScores(sessionId: string) {
    const session = sessions[sessionId];
    if (!session) return;

    // Reset scores
    session.participants.forEach(p => p.score = 0);

    // Sum votes
    session.paragraphs.forEach(p => {
        const paragraphScore = Object.values(p.votes).reduce((a, b) => a + b, 0);
        const author = session.participants.find(part => part.userId === p.userId);
        if (author) {
            author.score += paragraphScore;
        }
    });

    // Determine winner(s)
    const sorted = [...session.participants].sort((a, b) => b.score - a.score);
    const maxScore = sorted[0]?.score ?? 0;
    session.winnerUserId = sorted[0]?.userId || null; // Kept for legacy compatibility
    session.status = 'ended';
    session.phase = 'ended';

    // Record stats using fixed points (only if not already recorded)
    if (!session.statsRecorded) {
        session.statsRecorded = true;
        (async () => {
            try {
                for (const p of session.participants) {
                    const isWinner = p.score === maxScore && maxScore > 0;
                    const result = isWinner ? 'win' : 'loss';
                    const fixedPoints = isWinner ? 200 : 50;
                    await recordGameResult(p.userId, 'story_builder', result, fixedPoints);
                }
            } catch (err) {
                console.error('[StoryBuilder] Failed to record game stats:', err);
            }
        })();
    }

    broadcastState(sessionId);
}

function broadcastState(sessionId: string) {
    const session = sessions[sessionId];
    if (!session) return;
    const state = JSON.stringify({ type: 'state', payload: session });
    connections[sessionId]?.forEach(client => {
        if (client.readyState === 1) {
            client.send(state);
        }
    });
}

File: backend/app/domain/activities/story_builder.py
import asyncio
import json
import logging
import time
from typing import Dict, List, Optional, Set

from fastapi import WebSocket

from app.domain.activities import models, service
from app.domain.identity import audit
from app.domain.xp.service import XPService
from app.domain.xp.models import XPAction
from app.infra.auth import AuthenticatedUser

logger = logging.getLogger(__name__)

class StoryBuilderManager:
    def __init__(self):
        self.service = service.ActivitiesService()

    async def handle_socketio_action(self, session_id: str, user_id: str, action_type: str, payload: dict):
        # Ensure payload has userId for compatibility with existing logic methods
        payload["userId"] = user_id
        
        try:
            activity = await self.service._require_activity(session_id)
        except Exception:
            return

        if action_type == "join":
            await self._handle_join(session_id, activity, payload)
        elif action_type == "ready":
            await self._handle_ready(session_id, activity, payload)
        elif action_type == "submit_paragraph":
            await self._handle_submit_paragraph(session_id, activity, payload)
        elif action_type == "vote_paragraph":
            await self._handle_vote_paragraph(session_id, activity, payload)

    async def _handle_join(self, session_id: str, activity: models.Activity, payload: dict):
        user_id = payload.get("userId")
        if not user_id:
            return

        # Update participants list if needed (though activity.user_a/b are fixed)
        # Story Builder might support more than 2 users in future, but for now it's 2.
        
        await self._broadcast_state(session_id, activity)

    async def _handle_ready(self, session_id: str, activity: models.Activity, payload: dict):
        user_id = payload.get("userId")
        if not user_id:
            return
            
        story_meta = activity.meta.get("story", {})
        ready_map = story_meta.get("ready", {})
        ready_map[user_id] = True
        story_meta["ready"] = ready_map
        activity.meta["story"] = story_meta
        
        # Check if all participants are ready
        participants = [activity.user_a, activity.user_b]
        if all(ready_map.get(uid) for uid in participants):
            activity.state = "active" # or 'writing'
            # Initialize writing phase if not already
            if "paragraphs" not in story_meta:
                story_meta["paragraphs"] = []
                story_meta["turnIndex"] = 0
                story_meta["currentTurnUserId"] = activity.user_a # Start with user_a

            # Award XP for starting a game
            try:
                xp_svc = XPService()
                # Determine opponent for each player
                await xp_svc.award_xp(activity.user_a, XPAction.GAME_PLAYED, {"activity_id": activity.id, "target_id": activity.user_b, "game": activity.kind})
                await xp_svc.award_xp(activity.user_b, XPAction.GAME_PLAYED, {"activity_id": activity.id, "target_id": activity.user_a, "game": activity.kind})
            except Exception:
                logger.exception("Failed to award GAME_PLAYED XP for StoryBuilder")
        
        await self.service._persist(activity)
        await self._broadcast_state(session_id, activity)

    async def _handle_submit_paragraph(self, session_id: str, activity: models.Activity, payload: dict):
        user_id = payload.get("userId")
        text = payload.get("text")
        if not user_id or not text:
            return

        story_meta = activity.meta.get("story", {})
        current_turn_user = story_meta.get("currentTurnUserId")
        
        if user_id != current_turn_user:
            return # Not your turn

        paragraphs = story_meta.get("paragraphs", [])
        paragraphs.append({
            "userId": user_id,
            "text": text,
            "votes": {}
        })
        story_meta["paragraphs"] = paragraphs
        
        # Advance turn
        participants = [activity.user_a, activity.user_b]
        current_idx = participants.index(user_id)
        next_idx = (current_idx + 1) % len(participants)
        story_meta["currentTurnUserId"] = participants[next_idx]
        story_meta["turnIndex"] = story_meta.get("turnIndex", 0) + 1
        
        # Check if max turns reached
        max_turns = story_meta.get("turns", 6)
        if len(paragraphs) >= max_turns:
            activity.state = "voting" # Move to voting phase
            story_meta["phase"] = "voting"
        
        activity.meta["story"] = story_meta
        await self.service._persist(activity)
        await self._broadcast_state(session_id, activity)

    async def _handle_vote_paragraph(self, session_id: str, activity: models.Activity, payload: dict):
        user_id = payload.get("userId")
        paragraph_index = payload.get("paragraphIndex")
        score = payload.get("score")
        
        if user_id is None or paragraph_index is None or score is None:
            return

        story_meta = activity.meta.get("story", {})
        paragraphs = story_meta.get("paragraphs", [])
        
        if 0 <= paragraph_index < len(paragraphs):
            paragraph = paragraphs[paragraph_index]
            # Prevent voting on own paragraph
            if paragraph["userId"] == user_id:
                return
                
            if "votes" not in paragraph:
                paragraph["votes"] = {}
            paragraph["votes"][user_id] = score
            paragraphs[paragraph_index] = paragraph
            story_meta["paragraphs"] = paragraphs
            
            # Check if voting is complete
            # Every paragraph (except own) should be voted on by every other participant
            all_voted = True
            participants = [activity.user_a, activity.user_b]
            for idx, p in enumerate(paragraphs):
                for participant in participants:
                    if participant != p["userId"]:
                        if participant not in p.get("votes", {}):
                            all_voted = False
                            break
                if not all_voted:
                    break
            
            if all_voted:
                activity.state = "completed"
                story_meta["phase"] = "ended"
                
                # Each user's total score is the sum of points across ALL their paragraphs
                user_scores = {uid: 0.0 for uid in participants}
                for p in paragraphs:
                    pts = float(sum(p.get("votes", {}).values()))
                    user_scores[p["userId"]] += pts
                
                # Determine winner (handle ties)
                best_score = max(user_scores.values()) if user_scores else 0
                winner_id = None
                winners = [uid for uid, s in user_scores.items() if s == best_score and best_score > 0]
                if len(winners) == 1:
                    winner_id = winners[0]
                
                story_meta["winnerUserId"] = winner_id
                story_meta["scores"] = user_scores

                # Create a proper ScoreBoard object for the activity system
                scoreboard = models.ScoreBoard(activity.id)
                scoreboard.totals = {uid: float(s) for uid, s in user_scores.items()}
                # Populate per-round breakdown for secondary display
                for idx, p in enumerate(paragraphs):
                    scoreboard.add_score(idx + 1, p["userId"], float(sum(p.get("votes", {}).values())))
                
                activity.meta["scoreboard"] = scoreboard.to_payload()
                activity.meta["story"] = story_meta
                
                # Persist and record outcome
                await self.service._persist(activity)
                await self.service._record_leaderboard_outcome(activity, scoreboard)

                # Log activity completion for feed
                try:
                    # Log for the winner if exists, otherwise first participant
                    actor_id = winner_id or activity.user_a
                    await audit.log_event(
                        "activity_completed",
                        user_id=actor_id,
                        meta={
                            "activity_id": activity.id,
                            "kind": activity.kind,
                            "winner_id": winner_id
                        }
                    )
                except Exception:
                    logger.exception("Failed to log activity_completed for StoryBuilder")
            else:
                activity.meta["story"] = story_meta
                await self.service._persist(activity)
            
            await self._broadcast_state(session_id, activity)

    async def _broadcast_state(self, session_id: str, activity: models.Activity):
        from app.domain.activities.sockets import emit_activity_state
        
        story_meta = activity.meta.get("story", {})
        
        # Map activity state to frontend state
        status = "pending"
        if activity.state == "active":
            status = "writing"
        elif activity.state == "voting":
            status = "voting"
        elif activity.state == "completed":
            status = "ended"
            
        # Frontend expects 'participants' array with ready/joined status
        participants = []
        ready_map = story_meta.get("ready", {})
        scores = story_meta.get("scores", {})
        
        for uid in [activity.user_a, activity.user_b]:
            participants.append({
                "userId": uid,
                "joined": True, # Assume joined if in activity
                "ready": ready_map.get(uid, False),
                "score": scores.get(uid, 0)
            })

        state_payload = {
            "id": activity.id,
            "activityKey": "story_builder",
            "status": status,
            "phase": story_meta.get("phase", "lobby"),
            "lobbyReady": all(p["ready"] for p in participants),
            "creatorUserId": activity.user_a, # Assumption
            "participants": participants,
            "createdAt": int(activity.created_at.timestamp() * 1000),
            "paragraphs": story_meta.get("paragraphs", []),
            "maxParagraphsPerUser": 3, # derived from turns / 2
            "currentTurnUserId": story_meta.get("currentTurnUserId"),
            "turnOrder": [activity.user_a, activity.user_b],
            "turnIndex": story_meta.get("turnIndex", 0),
            "winnerUserId": story_meta.get("winnerUserId"),
            "connected": True
        }
        
        await emit_activity_state(session_id, state_payload)

manager = StoryBuilderManager()
